x-pack/plugins/fleet/server/services/files/client.ts

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Readable } from 'stream';
+
+import assert from 'assert';
+
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { Logger } from '@kbn/core/server';
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+
+import type { FileClient } from '@kbn/files-plugin/server';
+import { createEsFileClient, createFileHashTransform } from '@kbn/files-plugin/server';
+import type { File } from '@kbn/files-plugin/common';
+
+import { v4 as uuidV4 } from 'uuid';
+
+import moment from 'moment';
+
+import { getFileDataIndexName, getFileMetadataIndexName } from '../../../common';
+
+import { FleetFileNotFound, FleetFilesClientError } from '../../errors';
+
+import type {
+  FileCustomMeta,
+  FleetFile,
+  FleetFileClientInterface,
+  FleetFileType,
+  FleetFileUpdatableFields,
+  HapiReadableStream,
+  HostUploadedFileMetadata,
+} from './types';
+
+/**
+ * Public interface for interacting with Files stored in Fleet indexes. Service is consumed via
+ * the Fleet `Plugin#start()` interface - {@link FleetStartContract}
+ */
+export class FleetFilesClient implements FleetFileClientInterface {
```

Comment:
yeah, that could be another approach. I'll keep this implementation for now, but may adjust when I come back around and update it for the new index patterns.

Reply:
I thought more about this and I think I will leave it as is. Changing it to two different classes would cause some `TS` headaches when retrieving the client instance. The consumer of it would have to use type guards to narrow the type down in order to not get errors, and that (IMO) would create more burden on the consumers of it.. 

See here: https://www.typescriptlang.org/play?#code/FAYwNghgzlAEDCYCWBTAdgFwIKwN7FkNgHMUMAKASj1gF9h7hRIYFl0MAhWFADw3QATOIlSYc+IrACuAB0EQBVPPUYgA9migZYIAE4pFKUR1gBeWOQwBPWSgBcsAORYnsAD7POTyo5PiPNjEucwA+PAIiJAAzSxs7czMLFx8IqSkDDGk9NFg0FAB3II4sKgBuSMJGDLJs3Pyi-y5yhiZgAHp2pO6e7oBtXU0oaQBbJDRiWABdXqYNLR1wYPNdAyMm8idvSgrmYIA6OQUlHeAgA

Reply:
The reason I implemented restrictions on the client for files `from-host` is because I did not want to expose mutating methods  (`create`, `update`, `delete`) for files that were created by the Hosts. Those should not be manipulated by this client.  A kibana consumer of this service (IMO) should not have the ability to manipulate files that it creates. The other reason is that they may have different indexes where the data will be stored (Issue pending), thus why we needed something to (an argument) to indicate the file direction. 

Implementing them separate and exposing two methods out of the Fleet's server Plugin start services would be a way to avoid the type casting and/or type guards needed.

I'll consider splitting them up when this code is updated with the new index names if you are ok with that.

Reply:
thanks for the explanation, Paul! of course, let's see how it turns out.

a last idea, also for future consideration, if we want to have only one `FileClient`, it could return a `ToHostFile` object on `create()` which could contain the methods `update()` and `delete()`, if this makes sense